kernel/src/schema/compare.rs

```diff
@@ -165,15 +166,50 @@ impl SchemaComparison for DataType {
                 self_map.value_type().can_read_as(read_map.value_type())?;
             }
             (a, b) => {
-                // TODO: In the future, we will change this to support type widening.
-                // See: #623
-                require!(a == b, Error::TypeMismatch);
+                require!(types_compatible(a, b), Error::TypeMismatch);
             }
         };
         Ok(())
     }
 }
 
+/// Checks if a source type can be read as a target type, including type widening.
+///
+/// Type widening rules supported:
+/// - Integer widening: byte → short → int → long
+/// - Float widening: float → double
+fn types_compatible(source: &DataType, target: &DataType) -> bool {
```

Comment:
Why not a function `DataType.can_read_as(DataType)`?

Reply:
AH this is part of the impl. In  that case, why not inline this to `fn can_read_as(&self, read_type: &Self) -> SchemaComparisonResult`